cli/fossilize_replay.cpp

```diff
@@ -2866,36 +2893,50 @@ struct ThreadedReplayer : StateCreatorInterface
-							                enqueue_pipeline(parent.second.hash, parent.second.info,
-							                                 parent.second.pipeline,
-							                                 parent.second.index + hash_offset + start_index,
-							                                 PARENT_PIPELINE_MEMORY_CONTEXT);
-						                }
-					                }
+					                vector<DerivedInfo> ordered_parents;
+					                for (const auto &parent : parents)
+						                ordered_parents.push_back(parent.second);
 
-					                parents.clear();
+					                sort(begin(ordered_parents), end(ordered_parents), [&](const DerivedInfo &a, const DerivedInfo &b) -> bool {
+						                return a.depth > b.depth;
+					                });
 
 					                // We might be pulling in a parent pipeline from another memory context next iteration,
 					                // so we need to wait for all normal memory contexts.
 					                for (unsigned i = 0; i < NUM_PIPELINE_MEMORY_CONTEXTS; i++)
 						                sync_worker_memory_context(i);
+
+					                uint32_t prev_depth = 0xFFFFFFFF;
+					                for (auto &parent : ordered_parents)
+					                {
+						                if (!parent.info)
+							                continue;
+
+						                if (!derived_work_item_is_satisfied(parent, pipelines))
+							                continue;
+
+						                if (prev_depth != parent.depth) {
+							                prev_depth = parent.depth;
+							                sync_worker_memory_context(PARENT_PIPELINE_MEMORY_CONTEXT);
+						                }
+
+						                resolve_shader_modules(parent.info);
+						                resolve_pipelines(parent, pipelines);
+						                enqueue_pipeline(parent.hash, parent.info,
+						                                 parent.pipeline,
+						                                 parent.index + hash_offset + start_index,
+						                                 PARENT_PIPELINE_MEMORY_CONTEXT);
+					                }
 				                }});
 			}
 
 			work.push_back({ get_order_index(ENQUEUE_DERIVED_PIPELINES_OFFSET),
-			                 [this, &pipelines, derived, memory_index, hash_offset, start_index]() {
+			                 [this, &pipelines, derived, &parents, memory_index, hash_offset, start_index]() {
 				                 // Go over all pipelines. If there are no further dependencies to resolve, we can go ahead and queue them up.
 				                 // If an entry exists in pipelines, we have queued up that hash earlier, but it might not be done compiling yet.
 				                 auto itr = unstable_remove_if(begin(*derived), end(*derived), [&](const DerivedInfo &info) -> bool {
-					                 return derived_work_item_is_satisfied(info, pipelines);
+					                 return derived_work_item_is_satisfied(info, pipelines) || parents.count(info.hash) != 0;
```

Comment:
I don't understand this line.

Reply:
Nested pipeline libraries are already compiled in ENQUEUE_SHADER_MODULE_SECONDARY_OFFSET. This and the change below filter out pipelines that were already compiled without triggering the `pipelines were not compiled because parent pipelines do not exist.` warning. I can add a comment or change the condition for the warning to make it clear what's happening.